immich-app/immich
# fix: multiple file download fix for android

# Multiple Photo Download Fix on Mobile App (Android) - #14516

## Issue
When downloading photos on the mobile app (Android), multiple taps on the download button result in duplicate downloads of the same file. This creates unnecessary duplicate files with names like `foto.jpg`, `foto (1).jpg`, `foto (2).jpg`, etc.

## Root Cause
- No mechanism to prevent multiple download attempts while a download is in progress
- File existence checking based only on filename, not content
- Lack of download state tracking

## Changes Made

### Download Service ([download.service.dart]
- Added hash-based file verification using SHA-256
- Implemented file existence checking using content hash
- Added methods to calculate and verify file hashes
- Added logging for download status and skips
- Prevent redundant downloads of identical content

### Download Provider ([download.provider.dart]
- Added download state tracking mechanism
- Implemented prevention of concurrent downloads for the same asset
- Added user feedback through toast messages
- Enhanced error handling for failed downloads

### Dependencies
- Added `crypto: ^3.0.3` package for hash calculation

## Testing
1. Download a photo from the app
2. Attempt to download the same photo multiple times
3. Verify that:
   - Only one copy is downloaded
   - User receives appropriate feedback
   - No duplicate files are created
4. Test with various file types (JPG, HEIC, Live Photos)
5. Verify hash calculation works correctly for different file sizes


## Related Issues
- Fixes #14516
